src/fenics_constitutive/solver.py

```diff
@@ -0,0 +1,318 @@
+from __future__ import annotations
+
+import basix
+import dolfinx as df
+import numpy as np
+import ufl
+from petsc4py import PETSc
+
+from .interfaces import IncrSmallStrainModel
+from .maps import SubSpaceMap, build_subspace_map
+from .stress_strain import ufl_mandel_strain
+
+
+def build_history(
+    law: IncrSmallStrainModel, mesh: df.mesh.Mesh, q_degree: int
+) -> df.fem.Function | dict[str, df.fem.Function] | None:
+    """Build the history space and function(s) for the given law.
+
+    Args:
+        law: The constitutive law.
+        mesh: Either the full mesh for a homogenous domain or the submesh.
+        q_degree: The quadrature degree.
+
+    Returns:
+        The history function(s) for the given law.
+
+    """
+    match law.history_dim:
+        case int():
+            Qh = ufl.VectorElement(
+                "Quadrature",
+                mesh.ufl_cell(),
+                q_degree,
+                quad_scheme="default",
+                dim=law.history_dim,
+            )
+            history_space = df.fem.FunctionSpace(mesh, Qh)
+            history = df.fem.Function(history_space)
+        case None:
+            history = None
+        case dict():
+            history = {}
+            for key, value in law.history_dim.items():
+                if isinstance(value, int):
+                    Qh = ufl.VectorElement(
+                        "Quadrature",
+                        mesh.ufl_cell(),
+                        q_degree,
+                        quad_scheme="default",
+                        dim=value,
+                    )
+                elif isinstance(value, tuple):
+                    Qh = ufl.TensorElement(
+                        "Quadrature",
+                        mesh.ufl_cell(),
+                        q_degree,
+                        quad_scheme="default",
+                        shape=value,
+                    )
+                history_space = df.fem.FunctionSpace(mesh, Qh)
+                history[key] = df.fem.Function(history_space)
+    return history
+
+
+class IncrSmallStrainProblem(df.fem.petsc.NonlinearProblem):
+    def __init__(
+        self,
+        laws: list[tuple[IncrSmallStrainModel, np.ndarray]] | IncrSmallStrainModel,
+        u: df.fem.Function,
+        bcs: list[df.fem.DirichletBCMetaClass],
+        q_degree: int = 1,
+        form_compiler_options: dict | None = None,
+        jit_options: dict | None = None,
+    ):
+        mesh = u.function_space.mesh
+        map_c = mesh.topology.index_map(mesh.topology.dim)
+        num_cells = map_c.size_local + map_c.num_ghosts
```

Comment:
Did you run the tests in parallel as well? I was wondering if the ghost cells need to be included.
The integration points are local to each cell and thus only the cells that are really owned by the process need to be included?

Reply:
Not yet. Good idea though. Better to have it tested from the beginning. Can we just run `mpirun -n 2 pytest` or is it more complicated?

Reply:
I think you are right. Something like `mpirun -np 2 python3 -m pytest`
I just checked the [dolfinx ci](https://github.com/FEniCS/dolfinx/blob/main/.circleci/config.yml#L86C1-L90C41)

Reply:
https://pytest-mpi.readthedocs.io/en/latest/usage.html

Of course there is a package for it:D

Reply:
Forgot to answer the 2nd part of your question. I think, we may need both. I think `fem.Function.x` contains the ghost values and since it is recommended over `fem.Function.vector` for potential PETSc free builds, this might just be some computational overhead that we need to accept.

Reply:
Moved to new issue, since this is more complex https://github.com/BAMresearch/fenics-constitutive/issues/47